Sources/Plasma/Apps/plClient/Mac-Cocoa/PLSServerStatus.mm

```diff
@@ -0,0 +1,86 @@
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+Additional permissions under GNU GPL version 3 section 7
+
+If you modify this Program, or any covered work, by linking or
+combining it with any of RAD Game Tools Bink SDK, Autodesk 3ds Max SDK,
+NVIDIA PhysX SDK, Microsoft DirectX SDK, OpenSSL library, Independent
+JPEG Group JPEG library, Microsoft Windows Media SDK, or Apple QuickTime SDK
+(or a modified version of those libraries),
+containing parts covered by the terms of the Bink SDK EULA, 3ds Max EULA,
+PhysX SDK EULA, DirectX SDK EULA, OpenSSL and SSLeay licenses, IJG
+JPEG Library README, Windows Media SDK EULA, or QuickTime SDK EULA, the
+licensors of this Program grant you additional
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+#import "PLSServerStatus.h"
+#include <string_theory/string>
+#include "StringTheory_NSString.h"
+#include "plNetGameLib/plNetGameLib.h"
+
+@interface PLSServerStatus () <NSURLSessionDelegate>
+@property NSString* serverStatusString;
+@end
+
+@implementation PLSServerStatus
+
++ (id)sharedStatus
+{
+    static PLSServerStatus* shared = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        shared = [[self alloc] init];
+    });
+    return shared;
+}
+
+- (void)loadServerStatus
+{
+    NSString* urlString = [NSString stringWithSTString:GetServerStatusUrl()];
+    NSURL* url = [NSURL URLWithString:urlString];
+    NSURLSessionConfiguration* URLSessionConfiguration =
+        [NSURLSessionConfiguration ephemeralSessionConfiguration];
+    NSURLSession* session = [NSURLSession sessionWithConfiguration:URLSessionConfiguration
+                                                          delegate:self
+                                                     delegateQueue:NSOperationQueue.mainQueue];
+    NSURLSessionTask* statusTask = [session
+          dataTaskWithURL:url
+        completionHandler:^(NSData* _Nullable data, NSURLResponse* _Nullable response,
+                            NSError* _Nullable error) {
+            if (data) {
+                NSString* statusString = [[NSString alloc] initWithData:data
+                                                               encoding:NSUTF8StringEncoding];
+                self.serverStatusString = statusString;
+            }
+        }];
+    [statusTask resume];
```

Comment:
Is there a practical benefit to doing this with NSURLSession instead of cURL?

Reply:
In since this is Obj-C, I'd ask what the benefit is to doing it with cURL. :p

If this was ever moved to cross platform code - I could see using cURL.. But in since this is Obj-C, using C/C++ types just adds a lot of overhead.

Reply:
The Windows version was ported from WinHTTP to curl back in the day to make future cross-platform work easier, and @dpogue's Linux client naturally uses curl, so it would be good to use the same code on macOS for consistency. Of course, as with other things in the plClient code, the status message fetching isn't factored out yet to allow sharing it across the different platforms...

Reply:
Spent a few hours today looking at the cURL code. I have mixed feelings. None to do with cURL - mostly to do with that the Mac code would be divergent.

I'm not actually sure the cURL code in Windows is correct. The write function may be called several times as data is chunked - and the write function should continuous append to an existing buffer from that callback. PostMessage should not be called from the write callback. PostMessage should be called after curl_easy_perform (which is also seems to be?). Apple's FoundationNetworking library wraps that complexity - which is why that isn't handled with NSURLSessionTask.

The other issue is that you can't use a C callback to an Obj-C object instance (which is also a problem in C++.) The Win32 client is basically written in C style - so thats not an issue over there. On the Mac side I have some things to get cURL talking to my instance, and then get things adapted to NSString so the right UI messages can be sent.

So basically - it could be done but the cURL code would be rewritten. Which maybe gets us closer to a cross platform version. I dunno. The reason this is in NSURLSession is because I wanted to avoid the complexity something like cURL introduces. But if the cURL code needs to be written anyway that's not saving much.

Ideally - this should probably be a C++ method that takes a lambda, and all that complexity could be folded under that method. A lambda would also mean I could sidestep the C callback issues and pass an Obj-C instance into the lambda. The way it's wrapped can also be reshuffled - but this code is also intentionally within an Obj-C project with an NSString property because it needs to be part of the bindings system to forward messages to the UI. So somewhere this code is going to be wrapped in an Obj-C object (does not necessarily have to be a standalone object.)

**A few other non essential notes I made when looking this over**

One thing about Plasma is that it does not care what thread its callbacks get called on. This is because Win32 can be used from any thread. Cocoa UI can only be called from the main thread - so this creates some amount of busy work syncing Plasma onto the right thread. Not a big deal - just another thing I noted that would need to be dealt with for a cURL integration.

It's not directly related here, but also might help explain some architectural decisions here... It's usually good Obj-C manners to wrap any C++ code and not let it escape your class interface. Which I know is going to sound super weird to a bunch of C++ devs.

Obj-C++ is somewhat of a hack. Traditionally it's led to strange behaviors, and there are still a few features that don't work in Obj-C++. It's also sticky. You have to be Obj-C++ to import an Obj-C++ header. So Obj-C++ can cause issues in a code base. So - where possible - the Cocoa client treats C++ as a private implementation detail that is not to be exposed. That won't always be maintainable - but is still something the Plasma client tries to hold to.

The big thing that's incompatible with Obj-C++ is Swift. Swift cannot import any header that contains C++. So the Obj-C code attempts to provide a sane interface that is Obj-C pure for if Swift code needs to be dropped in. That's something more important for a SwiftUI client. An Obj-C header can't expose a C++ type if that class is to be carried to SwiftUI.

While that's been the architectural decision so far to maintain compatibility for Swift and SwiftUI - that might change. The next version of Xcode will include Swift support for C++. I have to investigate more to see if it's possible to lower our guard and let C++ be more present in the Obj-C headers.

Anyway - thats why there is an inclination to use Obj-C types. It doesn't directly affect cURL (Swift does support plain C types and headers.) But if you're wondering why Obj-C is avoiding using C++ types or String Theory in the headers - that's probably why.

Reply:
Following back up with this - my proposal would be to leave this as is for now, and make a change later in a future commit.